{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "1ecabdfc-4722-4f32-bcb9-c06336a6536d",
   "metadata": {},
   "outputs": [],
   "source": [
    "import subprocess\n",
    "import myvariant\n",
    "import os\n",
    "import sys\n",
    "import bisect\n",
    "import pyranges as pr\n",
    "from cyvcf2 import VCF, Writer\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from tqdm.notebook import tqdm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "13fbc0d4-a0ee-43e5-8fbc-9bb0d5551cb0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(185985, 6)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>dbGAP_id</th>\n",
       "      <th>dataset</th>\n",
       "      <th>file_name</th>\n",
       "      <th>extension</th>\n",
       "      <th>size_MB</th>\n",
       "      <th>path</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>phs000974-v6-c1</td>\n",
       "      <td>nih-nhlbi-topmed-freeze10-phs000974-v6-c1</td>\n",
       "      <td>Release_Notes.phs000974.TOPMed_WGS_Framingham....</td>\n",
       "      <td>pdf</td>\n",
       "      <td>0.449</td>\n",
       "      <td>/sbgenomics/project-files/Omics/nih-nhlbi-topm...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>phs000974-v6-c1</td>\n",
       "      <td>nih-nhlbi-topmed-freeze10-phs000974-v6-c1</td>\n",
       "      <td>Study_Report.phs000974.TOPMed_WGS_Framingham.v...</td>\n",
       "      <td>pdf</td>\n",
       "      <td>0.137</td>\n",
       "      <td>/sbgenomics/project-files/Omics/nih-nhlbi-topm...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>phs000974-v6-c1</td>\n",
       "      <td>nih-nhlbi-topmed-freeze10-phs000974-v6-c1</td>\n",
       "      <td>manifest_phs000974.TOPMed_WGS_Framingham.v6.p5...</td>\n",
       "      <td>pdf</td>\n",
       "      <td>0.047</td>\n",
       "      <td>/sbgenomics/project-files/Omics/nih-nhlbi-topm...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>phs000974-v6-c1</td>\n",
       "      <td>nih-nhlbi-topmed-freeze10-phs000974-v6-c1</td>\n",
       "      <td>phs000974.v6.pht004909.v5.TOPMed_FHS_Subject.d...</td>\n",
       "      <td>xml</td>\n",
       "      <td>0.001</td>\n",
       "      <td>/sbgenomics/project-files/Omics/nih-nhlbi-topm...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>phs000974-v6-c1</td>\n",
       "      <td>nih-nhlbi-topmed-freeze10-phs000974-v6-c1</td>\n",
       "      <td>phs000974.v6.pht004909.v5.p5.TOPMed_FHS_Subjec...</td>\n",
       "      <td>txt.gz</td>\n",
       "      <td>0.038</td>\n",
       "      <td>/sbgenomics/project-files/Omics/nih-nhlbi-topm...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          dbGAP_id                                    dataset  \\\n",
       "0  phs000974-v6-c1  nih-nhlbi-topmed-freeze10-phs000974-v6-c1   \n",
       "1  phs000974-v6-c1  nih-nhlbi-topmed-freeze10-phs000974-v6-c1   \n",
       "2  phs000974-v6-c1  nih-nhlbi-topmed-freeze10-phs000974-v6-c1   \n",
       "3  phs000974-v6-c1  nih-nhlbi-topmed-freeze10-phs000974-v6-c1   \n",
       "4  phs000974-v6-c1  nih-nhlbi-topmed-freeze10-phs000974-v6-c1   \n",
       "\n",
       "                                           file_name extension  size_MB  \\\n",
       "0  Release_Notes.phs000974.TOPMed_WGS_Framingham....       pdf    0.449   \n",
       "1  Study_Report.phs000974.TOPMed_WGS_Framingham.v...       pdf    0.137   \n",
       "2  manifest_phs000974.TOPMed_WGS_Framingham.v6.p5...       pdf    0.047   \n",
       "3  phs000974.v6.pht004909.v5.TOPMed_FHS_Subject.d...       xml    0.001   \n",
       "4  phs000974.v6.pht004909.v5.p5.TOPMed_FHS_Subjec...    txt.gz    0.038   \n",
       "\n",
       "                                                path  \n",
       "0  /sbgenomics/project-files/Omics/nih-nhlbi-topm...  \n",
       "1  /sbgenomics/project-files/Omics/nih-nhlbi-topm...  \n",
       "2  /sbgenomics/project-files/Omics/nih-nhlbi-topm...  \n",
       "3  /sbgenomics/project-files/Omics/nih-nhlbi-topm...  \n",
       "4  /sbgenomics/project-files/Omics/nih-nhlbi-topm...  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#all_Omics_files = pd.read_csv(f'/sbgenomics/workspace/Tables/all_Omics_file_index_202511.csv', index_col=0)\n",
    "all_Omics_files = pd.read_csv(f'/sbgenomics/workspace/Tables/all_Omics_file_index_202601.csv', index_col=0)\n",
    "print(all_Omics_files.shape)\n",
    "all_Omics_files.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "02308ba5-483c-4a9a-8965-9b483f3af04e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(24, 6)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>dbGAP_id</th>\n",
       "      <th>dataset</th>\n",
       "      <th>file_name</th>\n",
       "      <th>extension</th>\n",
       "      <th>size_MB</th>\n",
       "      <th>path</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3793</th>\n",
       "      <td>phs000974-v6-c2</td>\n",
       "      <td>nih-nhlbi-topmed-freeze10-phs000974-v6-c2</td>\n",
       "      <td>FRAMINGHAM_phs000974_TOPMed_WGS_freeze.10b.chr...</td>\n",
       "      <td>vcf.gz</td>\n",
       "      <td>1186.074</td>\n",
       "      <td>/sbgenomics/project-files/Omics/nih-nhlbi-topm...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3795</th>\n",
       "      <td>phs000974-v6-c2</td>\n",
       "      <td>nih-nhlbi-topmed-freeze10-phs000974-v6-c2</td>\n",
       "      <td>FRAMINGHAM_phs000974_TOPMed_WGS_freeze.10b.chr...</td>\n",
       "      <td>vcf.gz</td>\n",
       "      <td>710.615</td>\n",
       "      <td>/sbgenomics/project-files/Omics/nih-nhlbi-topm...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3797</th>\n",
       "      <td>phs000974-v6-c2</td>\n",
       "      <td>nih-nhlbi-topmed-freeze10-phs000974-v6-c2</td>\n",
       "      <td>FRAMINGHAM_phs000974_TOPMed_WGS_freeze.10b.chr...</td>\n",
       "      <td>vcf.gz</td>\n",
       "      <td>726.487</td>\n",
       "      <td>/sbgenomics/project-files/Omics/nih-nhlbi-topm...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3799</th>\n",
       "      <td>phs000974-v6-c2</td>\n",
       "      <td>nih-nhlbi-topmed-freeze10-phs000974-v6-c2</td>\n",
       "      <td>FRAMINGHAM_phs000974_TOPMed_WGS_freeze.10b.chr...</td>\n",
       "      <td>vcf.gz</td>\n",
       "      <td>697.103</td>\n",
       "      <td>/sbgenomics/project-files/Omics/nih-nhlbi-topm...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3801</th>\n",
       "      <td>phs000974-v6-c2</td>\n",
       "      <td>nih-nhlbi-topmed-freeze10-phs000974-v6-c2</td>\n",
       "      <td>FRAMINGHAM_phs000974_TOPMed_WGS_freeze.10b.chr...</td>\n",
       "      <td>vcf.gz</td>\n",
       "      <td>510.269</td>\n",
       "      <td>/sbgenomics/project-files/Omics/nih-nhlbi-topm...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             dbGAP_id                                    dataset  \\\n",
       "3793  phs000974-v6-c2  nih-nhlbi-topmed-freeze10-phs000974-v6-c2   \n",
       "3795  phs000974-v6-c2  nih-nhlbi-topmed-freeze10-phs000974-v6-c2   \n",
       "3797  phs000974-v6-c2  nih-nhlbi-topmed-freeze10-phs000974-v6-c2   \n",
       "3799  phs000974-v6-c2  nih-nhlbi-topmed-freeze10-phs000974-v6-c2   \n",
       "3801  phs000974-v6-c2  nih-nhlbi-topmed-freeze10-phs000974-v6-c2   \n",
       "\n",
       "                                              file_name extension   size_MB  \\\n",
       "3793  FRAMINGHAM_phs000974_TOPMed_WGS_freeze.10b.chr...    vcf.gz  1186.074   \n",
       "3795  FRAMINGHAM_phs000974_TOPMed_WGS_freeze.10b.chr...    vcf.gz   710.615   \n",
       "3797  FRAMINGHAM_phs000974_TOPMed_WGS_freeze.10b.chr...    vcf.gz   726.487   \n",
       "3799  FRAMINGHAM_phs000974_TOPMed_WGS_freeze.10b.chr...    vcf.gz   697.103   \n",
       "3801  FRAMINGHAM_phs000974_TOPMed_WGS_freeze.10b.chr...    vcf.gz   510.269   \n",
       "\n",
       "                                                   path  \n",
       "3793  /sbgenomics/project-files/Omics/nih-nhlbi-topm...  \n",
       "3795  /sbgenomics/project-files/Omics/nih-nhlbi-topm...  \n",
       "3797  /sbgenomics/project-files/Omics/nih-nhlbi-topm...  \n",
       "3799  /sbgenomics/project-files/Omics/nih-nhlbi-topm...  \n",
       "3801  /sbgenomics/project-files/Omics/nih-nhlbi-topm...  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#study = 'parent-CHS'\n",
    "#study = 'parent-MESA'\n",
    "#study = 'topmed-CHS'\n",
    "#study = 'topmed-JHS'\n",
    "#study = 'topmed-MESA'\n",
    "study = 'topmed-FHS'\n",
    "id = 'c2'\n",
    "select_files = all_Omics_files[all_Omics_files['dataset'] == f'nih-nhlbi-topmed-freeze10-phs000974-v6-c2']\n",
    "select_files = select_files[select_files['extension'] == 'vcf.gz']\n",
    "print(select_files.shape)\n",
    "select_files.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "e4feb8e5-fafb-467a-953c-725e60bafe3a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>11873</td>\n",
       "      <td>14409</td>\n",
       "      <td>DDX11L1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>14361</td>\n",
       "      <td>29370</td>\n",
       "      <td>WASH7P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>17368</td>\n",
       "      <td>17436</td>\n",
       "      <td>MIR6859-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>17368</td>\n",
       "      <td>17436</td>\n",
       "      <td>MIR6859-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>17368</td>\n",
       "      <td>17436</td>\n",
       "      <td>MIR6859-3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24904</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57184215</td>\n",
       "      <td>57197337</td>\n",
       "      <td>IL9R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73013</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57184215</td>\n",
       "      <td>57197337</td>\n",
       "      <td>IL9R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73695</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57201083</td>\n",
       "      <td>57203350</td>\n",
       "      <td>WASIR1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73696</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57212177</td>\n",
       "      <td>57214703</td>\n",
       "      <td>DDX11L16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73697</th>\n",
       "      <td>chrY_KZ208924v1_fix</td>\n",
       "      <td>4658</td>\n",
       "      <td>15725</td>\n",
       "      <td>TTTY13</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>84787 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     chrom     start       end       gene\n",
       "0                     chr1     11873     14409    DDX11L1\n",
       "1                     chr1     14361     29370     WASH7P\n",
       "2                     chr1     17368     17436  MIR6859-1\n",
       "3                     chr1     17368     17436  MIR6859-2\n",
       "4                     chr1     17368     17436  MIR6859-3\n",
       "...                    ...       ...       ...        ...\n",
       "24904                 chrY  57184215  57197337       IL9R\n",
       "73013                 chrY  57184215  57197337       IL9R\n",
       "73695                 chrY  57201083  57203350     WASIR1\n",
       "73696                 chrY  57212177  57214703   DDX11L16\n",
       "73697  chrY_KZ208924v1_fix      4658     15725     TTTY13\n",
       "\n",
       "[84787 rows x 4 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes_df = pd.read_csv(\"/sbgenomics/workspace/Tables/ref/hg38.refGene.genes.bed\", sep=\"\\t\", names=[\"chrom\", \"start\", \"end\", \"gene\"]).sort_values([\"chrom\", \"start\"])\n",
    "genes_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "286ef726-1a4d-400f-8be9-67a4998b2bab",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "d3a51dfa-4439-41ed-a67a-e14bc8da1f34",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "d8beab327bfd42b5ac0cb329e1c50a8b",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Files:   0%|          | 0/5 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "*********************1/24 FRAMINGHAM_phs000974_TOPMed_WGS_freeze.10b.chr1.hg38.c2*********************\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Lines   total/split/joined/realigned/skipped:\t3123650/0/0/0/0\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of samples: 883\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Step 3 (Python): 0it [00:00, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Final variants kept: 23295\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Output: 0it [00:00, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Input sequence provided is already in string format. No operation performed\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Final output (all SNPs): (23295, 890)\n",
      "Final output (remove non-idendified SNPs): (23295, 890)\n",
      "Final output (After dbSNP annotation and filtering): (7789, 890)\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "dbSNP annotation:   0%|          | 0/7789 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "*********************2/24 FRAMINGHAM_phs000974_TOPMed_WGS_freeze.10b.chr10.hg38.c2*********************\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Lines   total/split/joined/realigned/skipped:\t1951896/0/0/0/0\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of samples: 883\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Step 3 (Python): 0it [00:00, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Final variants kept: 8875\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Output: 0it [00:00, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Input sequence provided is already in string format. No operation performed\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Final output (all SNPs): (8875, 890)\n",
      "Final output (remove non-idendified SNPs): (8875, 890)\n",
      "Final output (After dbSNP annotation and filtering): (2720, 890)\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "dbSNP annotation:   0%|          | 0/2720 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "*********************3/24 FRAMINGHAM_phs000974_TOPMed_WGS_freeze.10b.chr11.hg38.c2*********************\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Lines   total/split/joined/realigned/skipped:\t1958687/0/0/0/0\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of samples: 883\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Step 3 (Python): 0it [00:00, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Final variants kept: 15184\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Output: 0it [00:00, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Input sequence provided is already in string format. No operation performed\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Final output (all SNPs): (15184, 890)\n",
      "Final output (remove non-idendified SNPs): (15184, 890)\n",
      "Final output (After dbSNP annotation and filtering): (5484, 890)\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "dbSNP annotation:   0%|          | 0/5484 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "*********************4/24 FRAMINGHAM_phs000974_TOPMed_WGS_freeze.10b.chr12.hg38.c2*********************\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Lines   total/split/joined/realigned/skipped:\t1866253/0/0/0/0\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of samples: 883\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Step 3 (Python): 0it [00:00, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Final variants kept: 11223\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Output: 0it [00:00, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Input sequence provided is already in string format. No operation performed\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Final output (all SNPs): (11223, 890)\n",
      "Final output (remove non-idendified SNPs): (11223, 890)\n",
      "Final output (After dbSNP annotation and filtering): (3962, 890)\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "dbSNP annotation:   0%|          | 0/3962 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "*********************5/24 FRAMINGHAM_phs000974_TOPMed_WGS_freeze.10b.chr13.hg38.c2*********************\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Lines   total/split/joined/realigned/skipped:\t1399978/0/0/0/0\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of samples: 883\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Step 3 (Python): 0it [00:00, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Final variants kept: 3780\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Output: 0it [00:00, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Input sequence provided is already in string format. No operation performed\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Final output (all SNPs): (3780, 890)\n",
      "Final output (remove non-idendified SNPs): (3780, 890)\n",
      "Final output (After dbSNP annotation and filtering): (1396, 890)\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "dbSNP annotation:   0%|          | 0/1396 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "## ======================================================\n",
    "## Integrating bcftools\n",
    "## ======================================================\n",
    "\n",
    "log_path = f\"/sbgenomics/workspace/Tables/VCF/{study}/{id}/output_1.txt\"\n",
    "os.makedirs(os.path.dirname(log_path), exist_ok=True)\n",
    "\n",
    "if os.path.exists(log_path):\n",
    "    os.remove(log_path)\n",
    "\n",
    "log_fh = open(log_path, \"w\")  ## append mode\n",
    "\n",
    "def fprint(*args, **kwargs):\n",
    "    print(*args, **kwargs)\n",
    "    print(*args, **kwargs, file=log_fh)\n",
    "    log_fh.flush()\n",
    "\n",
    "count = 0\n",
    "for file_id in tqdm(select_files.index[:], desc='Files'):\n",
    "    count += 1\n",
    "    fprint(f\"*********************{count}/{select_files.shape[0]} {select_files.loc[file_id, 'file_name'][:-7]}*********************\")\n",
    "    \n",
    "    path = select_files.loc[file_id, 'path']\n",
    "    if \",\" in path:\n",
    "        path = path.split(\",\")[0].strip().strip(\"'\").strip('\"')\n",
    "\n",
    "    out_dir = f\"/sbgenomics/workspace/Tables/VCF/{study}/{id}\"\n",
    "    \n",
    "    step1_vcf = f\"{out_dir}/step1_1.vcf.gz\"\n",
    "    step2_vcf = f\"{out_dir}/step2_1.vcf.gz\"\n",
    "    final_vcf = f\"{out_dir}/step3_1.vcf.gz\"\n",
    "\n",
    "    ## ======================================================\n",
    "    ## STEP 1 — Basic quality filtering (bcftools)\n",
    "    ## QUAL >= 30, FILTER == PASS\n",
    "    ## ======================================================\n",
    "    subprocess.run(\n",
    "        [\n",
    "            \"bcftools\", \"view\",\n",
    "            \"-i\", 'QUAL>=30 && FILTER=\"PASS\"',\n",
    "            \"-Oz\",\n",
    "            \"-o\", step1_vcf,\n",
    "            path\n",
    "        ],\n",
    "        check=True\n",
    "    )\n",
    "    subprocess.run([\"bcftools\", \"index\", \"-f\", step1_vcf], check=True)\n",
    "    \n",
    "    \n",
    "    ## ======================================================\n",
    "    ## STEP 2 — Normalize & split multiallelics (bcftools)\n",
    "    ## Equivalent to your Step 2a + 2b\n",
    "    ## ======================================================\n",
    "    subprocess.run(\n",
    "        [\n",
    "            \"bcftools\", \"norm\",\n",
    "            \"-m\", \"-any\",           # split multi-allelics\n",
    "            \"-Oz\",\n",
    "            \"-o\", step2_vcf,\n",
    "            step1_vcf\n",
    "        ],\n",
    "        check=True\n",
    "    )\n",
    "    subprocess.run([\"bcftools\", \"index\", \"-f\", step2_vcf], check=True)\n",
    "    \n",
    "    \n",
    "    ## ======================================================\n",
    "    ## STEP 3 — Annotation-based filtering (Python)\n",
    "    ## AF <= 0.01, missense / stop_gained, SNV only\n",
    "    ## ======================================================\n",
    "    vcf = VCF(step2_vcf)\n",
    "    writer = Writer(final_vcf, vcf)\n",
    "\n",
    "    fprint(\"Number of samples:\", len(vcf.samples))\n",
    "\n",
    "    kept = 0\n",
    "    for v in tqdm(vcf, leave=False, desc=\"Step 3 (Python)\"):\n",
    "        ## SNV only (safety check)\n",
    "        if len(v.REF) != 1 or any(len(a) != 1 for a in v.ALT):\n",
    "            continue\n",
    "\n",
    "        af = v.INFO.get(\"AF\")\n",
    "        if af is not None:\n",
    "            if isinstance(af, (list, tuple)):\n",
    "                af = min(af)\n",
    "            if af > 0.01:\n",
    "                continue\n",
    "\n",
    "        ann = v.INFO.get(\"ANN\", \"\")\n",
    "        if \"missense\" not in ann and \"stop_gained\" not in ann:\n",
    "            continue\n",
    "\n",
    "        writer.write_record(v)\n",
    "        kept += 1\n",
    "\n",
    "    writer.close()\n",
    "    vcf.close()\n",
    "    fprint(f\"Final variants kept: {kept}\")\n",
    "\n",
    "\n",
    "    \n",
    "    ## Output\n",
    "    vcf = VCF(f\"/sbgenomics/workspace/Tables/VCF/{study}/{id}/step3_1.vcf.gz\")\n",
    "    \n",
    "    gene_index = {}\n",
    "    for chrom, sub in genes_df.groupby(\"chrom\"):\n",
    "        gene_index[chrom] = {\n",
    "            \"starts\": sub.start.tolist(),\n",
    "            \"ends\": sub.end.tolist(),\n",
    "            \"genes\": sub.gene.tolist()\n",
    "        }\n",
    "    \n",
    "    def find_gene(chrom, pos0):\n",
    "        if chrom not in gene_index:\n",
    "            return None\n",
    "        starts = gene_index[chrom][\"starts\"]\n",
    "        ends = gene_index[chrom][\"ends\"]\n",
    "        genes = gene_index[chrom][\"genes\"]\n",
    "        i = bisect.bisect_right(starts, pos0) - 1\n",
    "        if i >= 0 and pos0 < ends[i]:\n",
    "            return genes[i]\n",
    "        return None\n",
    "    \n",
    "    def encode_gt(gt):\n",
    "        \"\"\"\n",
    "        gt is like [a, b, phased]\n",
    "        \"\"\"\n",
    "        a, b = gt[0], gt[1]\n",
    "        if a == -1 or b == -1:\n",
    "            return None\n",
    "        return a + b\n",
    "    \n",
    "    samples = vcf.samples\n",
    "    \n",
    "    records = []\n",
    "    for v in tqdm(vcf, leave=False, desc='Output'):\n",
    "        row = {\n",
    "            \"chrom\": v.CHROM,\n",
    "            \"pos\": v.POS,\n",
    "            \"ref\": v.REF,\n",
    "            \"alt\": v.ALT[0],\n",
    "            \"qual\": v.QUAL,\n",
    "            \"gene\": find_gene(v.CHROM, v.POS - 1),\n",
    "            #\"dp\": v.INFO.get(\"DP\"),\n",
    "            #\"af\": v.INFO.get(\"AF\"),\n",
    "            \"dbsnp_id\": v.ID if v.ID != \".\" else None\n",
    "        }\n",
    "        ## Add one column per sample\n",
    "        for sample, gt in zip(samples, v.genotypes):\n",
    "            row[sample] = encode_gt(gt)\n",
    "        records.append(row)\n",
    "    vcf.close()\n",
    "    \n",
    "    df = pd.DataFrame(records)\n",
    "    fprint(f\"Final output (all SNPs): {df.shape}\")\n",
    "    df = df[df.iloc[:,7:].sum(axis=1) >0]\n",
    "    fprint(f\"Final output (remove non-idendified SNPs): {df.shape}\")\n",
    "\n",
    "    ## dbSNP annotation\n",
    "    try:\n",
    "        ID_list = df['dbsnp_id'].values\n",
    "        selected_columns = [\n",
    "            \"_id\",\n",
    "            \"dbsnp.rsid\",\n",
    "            \"dbsnp.chrom\",\n",
    "            \"dbsnp.hg19.start\",\n",
    "            \"dbsnp.hg19.end\",\n",
    "            \"dbsnp.ref\",\n",
    "            \"dbsnp.alt\",\n",
    "            \"dbsnp.vartype\",\n",
    "            \"dbsnp.gene.symbol\",\n",
    "            \"snpeff.ann.effect\",\n",
    "            \"snpeff.ann.feature_id\",\n",
    "            \"snpeff.ann.feature_type\",\n",
    "            \"snpeff.ann.genename\",\n",
    "            \"snpeff.ann.gene_id\",\n",
    "            \"snpeff.ann.putative_impact\",\n",
    "            \"snpeff.ann.transcript_biotype\",\n",
    "            \"cadd.phred\",\n",
    "            \"cadd.rawscore\",\n",
    "            \"cadd.consequence\",\n",
    "            #\"cadd.fitcons\",\n",
    "            #\"cadd.gerp.s\"\n",
    "        ]\n",
    "        mv = myvariant.MyVariantInfo()\n",
    "        df_ids = mv.getvariants(ID_list, fields=\"dbsnp,snpeff,cadd,clinvar\", as_dataframe=True)\n",
    "        df_id_filter = df_ids[selected_columns]\n",
    "        keep_terms = ['missense_variant', 'stop_gained']\n",
    "        df_id_filter_filter = df_id_filter[df_id_filter['snpeff.ann.effect'].isin(keep_terms)]\n",
    "        df = df[df['dbsnp_id'].isin(set(df_id_filter_filter.index))]\n",
    "        fprint(f\"Final output (After dbSNP annotation and filtering): {df.shape}\")\n",
    "        rows = []\n",
    "        for i in tqdm(df.index, leave=False, desc=\"dbSNP annotation\"):\n",
    "            row_id = df.loc[i, 'dbsnp_id']\n",
    "            #data = df_id_filter_filter[df_id_filter_filter['query'] == row_id]\n",
    "            data = df_id_filter_filter.loc[[row_id]]\n",
    "            rows.append([data['snpeff.ann.effect'].values[0], data['dbsnp.gene.symbol'].values[0]])\n",
    "        rows = np.asarray(rows)\n",
    "        if \"var_type\" in df.columns:\n",
    "            df[\"var_type\"] = rows[:, 0]\n",
    "        else:\n",
    "            df.insert(7, \"var_type\", rows[:, 0])\n",
    "        if \"gene_dbsnp\" in df.columns:\n",
    "            df[\"gene_dbsnp\"] = rows[:, 1]\n",
    "        else:\n",
    "            df.insert(8, \"gene_dbsnp\", rows[:, 1])\n",
    "        \n",
    "        df.to_csv(f\"/sbgenomics/workspace/Tables/VCF/{study}/{id}/{select_files.loc[file_id, 'file_name'][:-7]}.csv\", index=True)\n",
    "    except:\n",
    "        fprint(\"No SNP identified!\")\n",
    "    \n",
    "    os.remove(f\"/sbgenomics/workspace/Tables/VCF/{study}/{id}/step1_1.vcf.gz\")\n",
    "    os.remove(f\"/sbgenomics/workspace/Tables/VCF/{study}/{id}/step2_1.vcf.gz\")\n",
    "    os.remove(f\"/sbgenomics/workspace/Tables/VCF/{study}/{id}/step3_1.vcf.gz\")\n",
    "    os.remove(f\"/sbgenomics/workspace/Tables/VCF/{study}/{id}/step1_1.vcf.gz.csi\")\n",
    "    os.remove(f\"/sbgenomics/workspace/Tables/VCF/{study}/{id}/step2_1.vcf.gz.csi\")\n",
    "\n",
    "log_fh.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "551d736c-371e-40db-9807-418b31163ba9",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
